### ISSUE TITLE
Overhaul savegame format and loading

### ISSUE THREAD

**Issue by Quintus** [Secretchronicles/TSC#380]
The savegame mechanism is currently quite complicated and error-prone. Just see the discussions in #376, #179, #186, and #354. The current level savegame system works like this:

* Saving
  * Each loaded level’s name is saved into the save file (think sublevels)
  * For each level, all _spawned_ objects are saved completely into the savegame, i.e. with all information required to recreate the `cSprite` subclass instance.
  * For each level, diffs are saved for the regular level objects, i.e. those have _not_ been spawned, but loaded from the level XML. The diff contains the difference in attributes between the original version in the level XML and the current version in the played game state.
* Loading
  * The XML file for the level itself is loaded for each level
  * All spawned objects are loaded by creating entirely new instances of the `cSprite` classes, which are then added to each level
  * All objects in each level are iterated. If the initial (X|Y) position of the object as encountered here matches the initial (X|Y) position of an object diff in the savegame, the diff is applied to the object (i.e. its attributes are changed accordingly).

The last step in the loading process is a mess, because the position match fails way to often. Also, handling spawned objects and regular objects differently results in lots of code duplication. There are two entirely separate hierarchies for loading objects, the `Load_From_Savegame()` virtual method hierarchy of `cSprite` and subclasses for only loading the diffs, and the `XmlAttributes` constructors of `cSprite` and subclasses, which are used to construct the spawned objects.

To remove all the code duplication, I suggest to unify the handling of spawned objects and regular objects so that all savegame loading always uses the `XmlAttributes` constructor and `Load_From_Savegame()` is removed completely. There’s no need to have two loading hierarchies with duplicate code. This requires that the regular objects are saved into the savegame XML with the same amount of information as the spawned objects, which is a **savegame format change** and hence breaks backward compatibility. The advantages are clear: No more code duplication, and even better, the hacky position checking goes away, because all objects are saved in their current state in the XML file.

We’ll have to discuss whether we want the static sprites to also save into the savegame XML. That would allow to load a savegame completely independant from the level file, so that there’s no fear that in an old savegame you suddenly saved over an abyss that was placed there in a newer version of the level.

Valete,
Quintus

**Comment by datahead8888** [comment]
Thanks for opening a ticket - I can't wait to have a reliable save system...

Just to check, will this maintain reverse compatibility?  If not, we will have to start the 3.0 branch now to use this.

**Comment by Quintus** [comment]
This save and reload system drives me mad. The constructors for objects saved in level XML do not cascade, but instead to some weird mix of their own `Init()` and a different parent class constructor.

I have arrived at a point where I would immediately rewrite the entire saving system from scratch and throw everything over board. Only backward compatibility blocks this, as it always does. The saving/loading code is crap, seriously. Making newer approaches work with the old format creates even more crappy code.

Valete,
Quintus

**Comment by datahead8888** [comment]
> I have arrived at a point where I would immediately rewrite the entire saving system from scratch and throw everything over board. Only backward compatibility blocks this, as it always does. The saving/loading code is crap, seriously. Making newer approaches work with the old format creates even more crappy code.

I'm guessing you read my thoughts in IRC on this.  Yes, the save system is a mess, and all of the reverse compatibility code is a mess (I know this because I did a change to allow .smclvl and .smcsav files to work).

I just wanted to suggest keeping the 2.X series for a while longer to allow a number of bug fixes and features (ie. alternate exits, some new enemies, the turtle boss, ice/fire blocks, etc) to go into this series.  This series is the only one that can potentially be forked into a new version of Secret Maryo Chronicles, and is the line of releases that fulfills the old SMC legacy.

SMC cannot be forked from the 3.X series (primarily because of mechanics changes but also to some extent reverse compatibility, a design tenet of SMC).  The moment we create the 3.0 branch, it will start to become more and more trouble to do merges into it.  I originally would have suggested planning for the 3.0 branch for a while to figure out what we want to do and creating it a little later.  I fully agree reverse compatibility must be dropped for the 3.X series and really like the idea of radically changing the mechanics in the 3.X series (so that we stop feeling like a Mario game).  I'm just interested in running the 2.X series for a while longer.

**Comment by brianvanderburg2** [comment]
I have an idea that can improve the save/load mechanism and still be backwards compatible for the time being if we decide to use the UID for saving in the future:
* Go ahead and save the UID in the savegame data along with pos_x and pos_y (cSprite::Save_to_Savegame should do this along with the rest)
* When loading, if UID is available in the save, try to use that to find the object. If not (or if the object was not found), use the pos_x/pos_y data.  Type should still match regardless if uid or position is used.

This is backward and forward compatible:
* Savegames saved without UID from an older version will use pos_x and pos_y when played with the newer version.
* Savegames saved without UID from an older version will use pos_x and pos_y when played with the older version.
* Savegames saved with UID from from a newer version will use UID when played with the newer version.
* Savegames saved with UID from a newer version will use pos_x and pos_y when played with the older version.

**Comment by Quintus** [comment]
Cf. #383 

----------

> Go ahead and save the UID in the savegame data along with pos_x and pos_y (cSprite::Save_to_Savegame should do this along with the rest)

I’d like to go one step even further. Currently, regular level objects and spawned objects are handled differently, which has no real reason in my opinion. I’d rather handle every object as a spawned one, with all the required information, and save the entire thing into the savegame file. Then there’s just one kind of things to cater for. I realise this approach doesn’t maintain backward compatibility, but it allows for cleaning up the code.

Vale,
Quintus

**Comment by sauer2** [comment]
@datahead8888 I don't think no one ever WILL want to fork the SMC code base.

**Comment by datahead8888** [comment]
>  I don't think no one ever WILL want to fork the SMC code base

Yes, this is what we established in the IRC meeting.  This will never happen without FluXy making a fork, and it seems questionable that will happen.  I'm not willing to spend much time on it by myself either.  I'm more committed to TSC than an SMC fork myself, too.
